### PR TITLE
ddl: implement intra-table DDL for SchemaTracker

### LINKED ISSUE #35933: light-weighted DDL and infoschema implementation for DM
## Enhancement

see https://github.com/pingcap/tiflow/issues/5334

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close #35933

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: ddl/schematracker/dm_tracker.go [hunk 1]
```diff
@@ -332,17 +344,695 @@ func (d SchemaTracker) DropView(ctx sessionctx.Context, stmt *ast.DropTableStmt)
...
+	indexPartSpecifications []*ast.IndexPartSpecification,
+	indexOption *ast.IndexOption,
+	ifNotExists bool,
+) error {
+	unique := keyType == ast.IndexKeyTypeUnique
+	tblInfo, err := d.TableByName(ti.Schema, ti.Name)
+	if err != nil {
+		return err
+	}
+	t := tables.MockTableFromMeta(tblInfo)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply ddl/schematracker/dm_tracker.go:367]: Checked that dm_tracker uses the following method of table.Table

- Meta
- Cols
- VisibleCols
- Indices

the related fields are constructed using the functions in normal path. I think it's OK.

File: ddl/schematracker/dm_tracker.go [hunk 2]
```diff
@@ -169,8 +172,17 @@ func (d SchemaTracker) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStm
 		return infoschema.ErrDatabaseNotExists.GenWithStackByArgs(ident.Schema)
 	}
 
+	strictSQLMode := ctx.GetSessionVars().StrictSQLMode
+	enableClusteredIndex := ctx.GetSessionVars().EnableClusteredIndex
+	// avoid to interfere session context in unit test
```
[the remark this answers is not in the document]
Comment (lance6716) [reply ddl/schematracker/dm_tracker.go]: done. https://github.com/pingcap/tidb/pull/36176/commits/f5c0a1692611710aa722b871dfaca6d48ac95b77

File: ddl/schematracker/dm_tracker.go [hunk 3]
```diff
@@ -332,17 +344,695 @@ func (d SchemaTracker) DropView(ctx sessionctx.Context, stmt *ast.DropTableStmt)
...
+func (d SchemaTracker) dropIndex(ctx sessionctx.Context, ti ast.Ident, indexName model.CIStr, ifExists bool) error {
+	tblInfo, err := d.TableByName(ti.Schema, ti.Name)
+	if err != nil {
+		return infoschema.ErrTableNotExists.GenWithStackByArgs(ti.Schema, ti.Name)
+	}
+	t := tables.MockTableFromMeta(tblInfo)
+
+	indexInfo := tblInfo.FindIndexByName(indexName.L)
+
+	_, err = ddl.CheckIsDropPrimaryKey(indexName, indexInfo, t)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply ddl/schematracker/dm_tracker.go]: removed

File: ddl/schematracker/dm_tracker.go [hunk 4]
```diff
@@ -332,17 +344,695 @@ func (d SchemaTracker) DropView(ctx sessionctx.Context, stmt *ast.DropTableStmt)
...
+
+	// do drop column
+	tblInfo.MoveColumnInfo(colInfo.Offset, len(tblInfo.Columns)-1)
+	tblInfo.Columns = tblInfo.Columns[:len(tblInfo.Columns)-1]
+
+	newIndices := make([]*model.IndexInfo, 0, len(tblInfo.Indices))
+	for _, idx := range tblInfo.Indices {
+		var i int
+		found = false
+		for i = range idx.Columns {
```
[the remark this answers is not in the document]
Comment (lance6716) [reply ddl/schematracker/dm_tracker.go]: the function has been changed to 

```
// FindColumnInIndexCols finds column and its slice offset in index. When not found, returned slice offset is -1 and
// *IndexColumn is nil.
func FindColumnInIndexCols(c string, cols []*model.IndexColumn) (int, *model.IndexColumn)
```

[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/schematracker/dm_tracker.go]: Why did we add a return value? Could we just check if it（`*model.IndexColumn`） is nil?

File: ddl/db_integration_test.go [hunk 5]
```diff
@@ -360,8 +399,13 @@ func TestIssue2858And2717(t *testing.T) {
 }
 
 func TestIssue4432(t *testing.T) {
-	store, clean := testkit.CreateMockStore(t)
+	store, dom, clean := testkit.CreateMockStoreAndDomain(t)
 	defer clean()
+
+	ddlChecker := schematracker.NewChecker(dom.DDL())
+	dom.SetDDL(ddlChecker)
+	ddlChecker.CreateTestDB()
```
Comment (zimulala) [inline ddl/db_integration_test.go:407]: Does this result in normal DDL operations not being tested?

File: ddl/schematracker/dm_tracker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lance6716) [reply ddl/schematracker/dm_tracker.go]: caller need to delete i-th column in the slice.

[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/schematracker/dm_tracker.go]: Got it.

File: ddl/db_integration_test.go (hunk 5 above)
Comment (lance6716) [reply ddl/db_integration_test.go:407]: ddlChecker will call DDL API for dom.DDL() and SchemaTracker both, so the normal DDL code path is also covered

**Comment by lance6716** [comment]
PTAL @wjhuang2016 @zimulala

File: ddl/db_integration_test.go (hunk 5 above)
Comment (zimulala) [reply ddl/db_integration_test.go:407]: Got it.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**